Reformatted BTD fields

RevathiJambunathan opened the pull request:
In this PR, back-transformed diagnostics is reformatted to use the new diagnostics framework.
The field buffer data is flushed out using the customized format from the existing BTD. These functions in the reformatted BTD are named with a prefix `TMP_`, since this is a temporary fix, until support for plotfile/OpenPMD format is added. (Issue #1167 )

An automated test for comparing the sliced-BTD with the full BTD snapshot is included for the reformatted BTD as WarpX-tests.ini titled "Reformatted_BTD_ReducedSliceDiag".

Coarsening ratio of (1 1 1) is only supported at this time, since CoarsenIO:: does not fill ghost cells of the destination cell-centered MultiFab. This ghost data is required when averaging-down the field-data to a single-level for BTD. 
This will be fixed in CoarsenIO in a follow-up PR and then the limitation on coarsening ratio will be removed. 

Only support for fields is included. Particle-buffers and their Lorentz-transform will be added in a future PR. (Issue #1166 )


The input parameters to be added to use the new BTD is as follows
```

diagnostics.diags_names = diag1 diag2
diag1.period = 10
diag1.diag_type = Full
diag2.diag_type = BackTransformed
diag2.do_back_transformed_fields = 1
diag2.num_snapshots_lab = 2
diag2.dt_snapshots_lab = 1.8679589331096515e-13
diag2.fields_to_plot = Ex Ey Ez Bx By Bz jx jy jz rho
```
Support for any user-defined `diag_lo` and `diag_hi`is also included.

I have verified the implementation by performing a 2D test-case in Examples/Modules/RigidInjection 
The test for back-transformed slice or line data is also performed and verified for accuracy.
![Screenshot from 2020-07-08 21-19-34](https://user-images.githubusercontent.com/41089244/86996535-cbe49d80-c160-11ea-8a5a-8f24c85c1fd2.png)

Input file for the RigidInjection test:

[inputs_2d_BoostedFrame_reformatted.txt](https://github.com/ECP-WarpX/WarpX/files/4949975/inputs_2d_BoostedFrame_reformatted.txt)

Analysis script (renamed the extension of this .py script to .txt since github dint allow upload of file with .py.)
[verify_reformatted_BTD.txt](https://github.com/ECP-WarpX/WarpX/files/4949991/verify_reformatted_BTD.txt)

atmyers commented on Examples/Modules/boosted_diags/inputs_3d_slice_reformatted:
```diff
@@ -0,0 +1,123 @@
+warpx.zmax_plasma_to_compute_max_step = 0.0031
```
Do we need both this and `inputs_3d_slice`?

RevathiJambunathan replied:
We could remove the old one actually. The old one tests the old implementation and new tests the new implementation. Will do whatever you suggest :)

atmyers replied:
I don't think it makes a big difference but IMO there's slightly less chance of confusion if there's only one inputs file.

atmyers commented on Source/Diagnostics/BTDiagnostics.H:
```diff
@@ -14,47 +14,88 @@ private:
-    // Currently this is an empty function. When OpenPMD/Plotfile is supported
-    // this function can be moved to the base class Diagnostics.
-    void Flush (int i_buffer) override {}
+    // Currently, a temporory customized output format for the buffer
```
spelling : temporary

RevathiJambunathan replied:
Thank you for catching this. I fixed it now.

atmyers commented on Source/Evolve/WarpXEvolve.cpp:
```diff
@@ -217,19 +225,18 @@ WarpX::Evolve (int numsteps)
-        for (int i = 0; i <= max_level; ++i) {
-            t_new[i] = cur_time;
-        }
+        //// sync up time
```
Do we still need this?

RevathiJambunathan replied:
I kept this one to be consistent with development branch and removed the previous one since it was added to compare the old and new BTD implementation.

atmyers commented on Regression/WarpX-tests.ini:
```diff
@@ -163,6 +163,25 @@ aux1File = Tools/PostProcessing/read_raw_data.py
 analysisRoutine = Examples/Modules/boosted_diags/analysis_3Dbacktransformed_diag.py
 tolerance = 1.e-14
 
+[Reformatted_BTD_ReducedSliceDiag]
```
If we turn this test off, maybe we should create an issue to keep track of the fact that we need to further split up the tests into more CI runners.

atmyers commented:
Looks like there is a conflict to be fixed.

RevathiJambunathan commented:
Thank you for reviewing @atmyers .
I just update the branch with development and fixed the conflict. Also addressed some unused parameter warnings from one of the Diagnostic functions added in this PR, in c33e40c

RevathiJambunathan commented:
I have removed the CI-test for the reformatted BTD slices and opened Issue #1234 to ensure the test is added when the time-out issue is resolved. Thank you for the suggestion @atmyers

MaxThevenet commented on Source/Diagnostics/BTDiagnostics.H:
```diff
@@ -14,47 +14,88 @@ private:
     /** Read relevant parameters for BTD */
     void ReadParameters ();
     /** Flush m_mf_output and particles to file. */
```
```suggestion
    /** \brief Flush m_mf_output and particles to file.
```

MaxThevenet commented on Source/Diagnostics/BTDiagnostics.H:
```diff
@@ -14,47 +14,88 @@ private:
-    // Currently this is an empty function. When OpenPMD/Plotfile is supported
-    // this function can be moved to the base class Diagnostics.
-    void Flush (int /*i_buffer*/) override {}
+    // Currently, a temporary customized output format for the buffer
+    // data is implemented and called in this function.
+    // After implementing OpenPMD and plotfile formats, this function
+    // can be defined and implemented in Diagnostics.H, such that,
+    // the function call to flush out buffer data for
+    // FullDiagnostics and BTDiagnostics is the same
```
I think it's alright to put this in the Doxygen doc, but that's just a suggestion
```suggestion
     * Currently, a temporary customized output format for the buffer
     * data is implemented and called in this function.
     * After implementing OpenPMD and plotfile formats, this function
     * can be defined and implemented in Diagnostics.H, such that,
     * the function call to flush out buffer data for
     * FullDiagnostics and BTDiagnostics is the same */
```

MaxThevenet commented on Source/Diagnostics/BTDiagnostics.H:
```diff
@@ -14,47 +14,88 @@ private:
      */
     void InitializeFieldFunctors (int lev) override;
-    /** This function initialized and allocates particle buffers for all the snapshots.
+    /** This function allocates and initializes particle buffers for all the snapshots.
      *  The particle containers required for this must be added to populate this function.
      */
```
Do I understand this well?
```suggestion
    /** This function allocates and initializes particle buffers for all the snapshots.
     * This is currently an empty function:
     * The particle containers required for this must be added to populate this function.
     */
```

RevathiJambunathan replied:
YEs. it is correct that the function is currently empty and particle containers are required to populate this function

MaxThevenet commented on Source/Diagnostics/BTDiagnostics.H:
```diff
@@ -14,47 +14,88 @@ private:
+     * \param[in] step current time step, return true for any step >= 0
+     * \param[in] force_flush if true, return true for any step
+     */
     bool DoComputeAndPack (int step, bool force_flush=false) override;
```
Would sth like that work?
```suggestion
    bool DoComputeAndPack (int step, bool force_flush=false) {return step>=0;} override;
```

MaxThevenet commented on Source/Diagnostics/BTDiagnostics.cpp:
```diff
@@ -153,17 +163,35 @@ bool
 BTDiagnostics::DoComputeAndPack (int step, bool force_flush)
 {
     // always set to true for BTDiagnostics since back-transform buffers are potentially
-    // computed and packed every timstep.
-    return true;
+    // computed and packed every timstep, except at initialization when step == -1.
+    if (step>=0) {
+        return true;
+    }
+    return false;
 }
```
If the short version in the .H file works, this could be removed.

RevathiJambunathan replied:
It was leading to compiler error. I reverted the implementation to cpp, but with return(step>=0)

RevathiJambunathan commented:
Thank you @MaxThevenet for reviewing the PR and for the suggestions.

RevathiJambunathan commented:
@MaxThevenet to ensure that the suggested addition to documentation is made, I have included it in the re-organize diagnostics project in the "to-do" column - so I can keep track

MaxThevenet commented:
Thanks! I'll merge when tests pass.

MaxThevenet commented:
:tada:
